### PR TITLE
14100 improve navigation creation

### LINKED ISSUE #14100: [Improvement]: Make adding main navigation items easier
### Improvement description

Currently the navigation items list is created via a hardcoded twig array before scripts are loaded.

```twig
 {% set navigations = {
      'file': {'label': 'file', 'icon': 'file'},
      'extras': {'label': 'tools', 'icon': 'build'},
      'marketing': {'label': 'marketing', 'icon': 'bar_chart'},
      'settings': {'label': 'settings', 'icon': 'settings'},
      'ecommerce': {'label': 'bundle_ecommerce_mainmenu', 'icon': 'shopping_cart', 'extraAttr': 'style="display: none;"'},
      'search': {'label': 'search', 'icon': 'search'}
  } %}
```

With the bundle refactoring a priority was introduced and the pimcoreReady event is abandoned and a new event preMenuBuild was introduced for creating submenus. 
There should be an easier and a more native way to add main menu items than via the pimcoreReady event and DOM manipulation.

### PR DESCRIPTION
**Pull request by mattamon**
## Changes in this pull request  
Resolves #14100 

## Additional info  
Removing the hardcoded html markup for the main navigation.
Navigation is now built dynamically after the preMenuBuild event
All main menu items now get a priority like every submenu item if the priority is not specified. Standard gap is 10 like for all submenus.
Main menu items now have an iconCls property instead of using an img tag for the icon.

Adding a new main menu item should now look like the following:
```js
    menu.file = {
          label: t('myBundleLabel'), // set your label here, will be shown as tooltip
          iconCls: 'pimcore_main_nav_icon_myIcon', // set full icon name here
          priority: 42, // define the position where you menu should be shown. Core menu items will leave a gap of 10 custom main menu items
          items: items, //if your main menu has subitems please see Adding Custom Submenus To ExistingNavigation Items 
          shadow: false,
          handler: this.openMyBundle, // defining a handler will override the standard "showSubMenu" functionality, use in combination with "noSubmenus"
          noSubmenus: true, // if there are no submenus set to true otherwise menu won't show up
          cls: "pimcore_navigation_flyout", // use pimcore_navigation_flyout if you have subitems
     };
```

CMF PR: https://github.com/pimcore/customer-data-framework/pull/406
Datahub PR: https://github.com/pimcore/data-hub/pull/678
E-Commerce will be extracted to it's own repository

### CODE REVIEW

File: bundles/AdminBundle/public/js/pimcore/helpers.js [hunk 1]
```diff
@@ -3235,3 +3239,76 @@ pimcore.helpers.buildMenu = function(items) {
...
+                cls: 'pimcore_menu_item pimcore_menu_needs_children',
+                html: '<div id="menuitem-' + key + '-iconEl" data-ref="iconEl" class="x-menu-item-main-icon x-menu-item-icon pimcore_main_nav_icon_' + sortedMenu[key]['iconCls'] + '"></div>',
+                'data-menu-tooltip': t(sortedMenu[key]['label'])
+            }
+            if(sortedMenu[key]['style']) {
+                li.style = sortedMenu[key]['style'];
+            }
+            dh.append(ul, li)
+        }
+    })
```
Comment (jdreesen) [inline bundles/AdminBundle/public/js/pimcore/helpers.js]: We don't really need the intermediate variable, do we?

```suggestion
    const dh = Ext.DomHelper;
    const ul = Ext.get("pimcore_navigation_ul");
    const menuPrefix = 'pimcore_menu_';
    
    // sorting must be done manually here.
    Object.keys(menu).sort((a, b) => {
        // a and b are the keys like file, extras e.t.c
        // we use the keys to get the menu object themselfes with the priorities
        return pimcore.helpers.priorityCompare(menu[a], menu[b]);
    }).filter(key => {
        // the notifications are excluded from this
        return !menu[key]['exclude'];
    }).forEach(key => {
        const li = {
            id: menuPrefix + key,
            tag: 'li',
            cls: 'pimcore_menu_item pimcore_menu_needs_children',
            html: '<div id="menuitem-' + key + '-iconEl" data-ref="iconEl" class="x-menu-item-main-icon x-menu-item-icon pimcore_main_nav_icon_' + menu[key]['iconCls'] + '"></div>',
            'data-menu-tooltip': t(menu[key]['label'])
        }
        if(menu[key]['style']) {
            li.style = menu[key]['style'];
        }
        dh.append(ul, li)
    })
```

Comment (mattamon) [reply bundles/AdminBundle/public/js/pimcore/helpers.js]: Good catch! We indeed don't need it! Thanks!

File: bundles/AdminBundle/public/extjs/css/PimcoreApp-all_1.css [hunk 2]
```diff
@@ -14354,6 +14354,18 @@ a.x-color-picker-item:hover {
     margin-right: 41px;
 }
 
+.x-menu-item-main-icon {
```
Comment (fashxp) [inline bundles/AdminBundle/public/extjs/css/PimcoreApp-all_1.css]: is that really the correct place for this - as this file should contain ExtJS default styles only and would be overwritten when updating ExtJs?

File: bundles/AdminBundle/public/js/pimcore/helpers.js [hunk 3]
```diff
@@ -3235,3 +3239,70 @@ pimcore.helpers.buildMenu = function(items) {
...
+        // the notifications are excluded from this
+        return !menu[key]['exclude'];
+    }).forEach(key => {
+        const li = {
+            id: menuPrefix + key,
+            tag: 'li',
+            cls: 'pimcore_menu_item pimcore_menu_needs_children',
+            html: '<div id="menuitem-' + key + '-iconEl" data-ref="iconEl" class="x-menu-item-main-icon x-menu-item-icon pimcore_main_nav_icon_' + menu[key]['iconCls'] + '"></div>',
+            'data-menu-tooltip': t(menu[key]['label'])
+        }
```
Comment (fashxp) [inline bundles/AdminBundle/public/js/pimcore/helpers.js]:
```suggestion
        };
```

File: bundles/AdminBundle/public/js/pimcore/helpers.js [hunk 4]
```diff
@@ -3235,3 +3239,70 @@ pimcore.helpers.buildMenu = function(items) {
...
+        if(menu[key]['style']) {
+            li.style = menu[key]['style'];
+        }
+        dh.append(ul, li)
```
Comment (fashxp) [inline bundles/AdminBundle/public/js/pimcore/helpers.js]:
```suggestion
        dh.append(ul, li);
```

File: bundles/AdminBundle/public/js/pimcore/helpers.js [hunk 5]
```diff
@@ -3235,3 +3239,70 @@ pimcore.helpers.buildMenu = function(items) {
...
+    // add the maintenance at last
+    dh.append(ul,
+        {
+            id: menuPrefix + 'maintenance',
+            tag: 'li',
+            cls: 'pimcore_menu_item',
+            style: 'display:none;',
+            'data-menu-tooltip': t('deactivate_maintenance')
+        }
+    )
```
Comment (fashxp) [inline bundles/AdminBundle/public/js/pimcore/helpers.js]:
```suggestion
    );
```

**Review by fashxp — CHANGES_REQUESTED** [review]

**Comment by dpfaffenbauer** [comment]
why not something like this? https://github.com/coreshop/MenuBundle.

Control it completely from the PHP side

**Review by fashxp — CHANGES_REQUESTED** [review]
we should upgrade notes and docs

**Comment by mattamon** [comment]
@dpfaffenbauer thanks for the input, we gonna discuss it in the next meeting :) maybe something for pimcore 12?

**Comment by dpfaffenbauer** [comment]
@mattamon sure, we could take it from coreshop and move it to pimcore.

**Comment by mattamon** [comment]
@dpfaffenbauer we discussed this and were in unison, that we will consider it for pimcore 12

**Comment by jdreesen** [comment]
So Pimcore 12 will come soon after Pimcore 11?

**Review by fashxp — APPROVED** [review]

**Pull request merged**